File: src/diff150/RedAndBlack.py


def dfs(x, y, mx, my):
    global tile
    global cnt
    if (x < 0) or (x > mx) or (y < 0) or (y > my):
        return
    if tile[x][y] == '#':
        return
    cnt += 1
    tile[x][y] = '#'
    dfs(x+1, y, mx, my)
    dfs(x-1, y, mx, my)
    dfs(x, y+1, mx, my)
    dfs(x, y-1, mx, my)


if __name__ == '__main__':
    global tile
    global cnt
    W, H = map(int, input().split())
    tile = [input() for i in range(H)]
    for i in range(H):
        sx, sy = tile[i].find('@'), i
        if sx > 0:
            break
    dfs(sy, sx, H, W)
    print(cnt)

File: src/diff150/RedAndBlack.cpp
#include "bits/stdc++.h"
using std::cin;
using std::cout;

int cnt = 0;
char map[21][21];

void dfs(int x, int y, int mx, int my){
    if((x < 0) || (x >= mx)){
        return;
    }
    if((y < 0) || (y >= my)){
        return;
    }
    if(map[x][y] == '#'){
        return;
    }
    map[x][y] = '#';
    cnt++;
    dfs(x+1, y, mx, my);
    dfs(x-1, y, mx, my);
    dfs(x, y+1, mx, my);
    dfs(x, y-1, mx, my);
}

int main(){
    int W, H, sx, sy;
    cin >> W >> H;

    while(W != 0 && H != 0){
        for(int i = 0; i < H; i++){
            scanf("%20s", map[i]);
            if(std::string(map[i]).find('@') != std::string::npos){
                sx = i;
                sy = std::string(map[i]).find('@');
            }
        }
        dfs(sx, sy, H, W);
        cout << cnt << std::endl;

        cin >> W >> H;
        cnt = 0;
        
    }
}